### DISCUSSION TITLE
Gmail mockup in unpoly

### DISCUSSION THREAD

**Discussion by josephernest** [unpoly/unpoly#406] [General]
The documentation paragraph "Leveraging layers" on https://unpoly.com/tutorial with Gmail example is very interesting for someone arriving to unpoly. 

For the tutorial purpose, it would be super interesting (and good for wider adoption!) to have a small Gmail mockup example made with unpoly + PHP (or any other server side lang) that would mimic `/messages`, `/messages/new`, `/contacts?query=...`.
This would be a good example for the layer system.

I'm thinking about details that I wouldn't know how to solve:

* `/messages` => click on New => URL changes to `/messages/new` => ok for this part, if the user refreshes the browser, it seems ok too
* Then you enter a contact, you now have 3 layers (Messages, New modal, and contact popup), the URL changes to `/contacts?query=...`. Then what happens if the user "refreshes" the browser (F5)? How would it go? `/contacts?query` will generate a full HTML page, as usual, but then the fact it was embedded in Message > New is lost. How do you handle this?

Thanks!

**Comment by MattApril** [comment]
I'm also curious to know if there is any built-in mechanism for restoring the layer states. I don't believe there is, but I would be curious if/how anyone else is doing it.

Personally, I just make sure each layer can function as its own individual page and I do not attempt to track or restore the layer states. If the layer doesn't make sense as its own page, and/or I don't want users to access it as its own page, I just don't update the URL. So for your example, I would change the URL to /messages/new, but I would not change the URL to /contacts when opening the 3rd layer. That way, if the user refreshes the page or copies the URL they will just see the new message page, and they would have to manually re-open the contact dialog.

**Comment by triskweline** [comment]
> For the tutorial purpose, it would be super interesting (and good for wider adoption!) to have a small Gmail mockup example made with unpoly + PHP (or any other server side lang) that would mimic /messages, /messages/new, /contacts?query=....
This would be a good example for the layer system.

There is a demo application on <https://demo.unpoly.com/> that use nested layers extensively. Every screen links to the source code on GitHub.
